syntax = "proto3";

package wms.go.service.batching.split.system.v1;
option go_package = "gitlab.ozon.ru/wms/go/service/batching/split/pkg/system_v1";

import "google/api/annotations.proto";
import "github.com/envoyproxy/protoc-gen-validate/validate/validate.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";


service SystemV1 {
  rpc GetFeatureFlagList (GetFeatureFlagListRequest) returns (GetFeatureFlagListResponse) {
      option (google.api.http) = {
          post: "/v1/system/get-feature-flag-list"
          body: "*"
      };
  }

  rpc LinkFeatureFlag (LinkFeatureFlagRequest) returns (LinkFeatureFlagResponse) {
      option (google.api.http) = {
          post: "/v1/system/link-feature-flag"
          body: "*"
      };
  }

  rpc UnlinkFeatureFlag (UnlinkFeatureFlagRequest) returns (UnlinkFeatureFlagResponse) {
      option (google.api.http) = {
          post: "/v1/system/unlink-feature-flag"
          body: "*"
      };
  }

  rpc V1GetSplitRestrictions (V1GetSplitRestrictionsRequest) returns (V1GetSplitRestrictionsResponse);
  rpc V1CreateSplitRestrictions (V1CreateSplitRestrictionsRequest) returns (V1CreateSplitRestrictionsResponse);
  rpc V1UpdateSplitRestrictions (V1UpdateSplitRestrictionsRequest) returns (V1UpdateSplitRestrictionsResponse);
}

message GetFeatureFlagListRequest {
  google.protobuf.Int64Value warehouse_id = 1;
}

message GetFeatureFlagListResponse {
  message FeatureFlag {
      int64 id = 1;
      string alias = 2;
      string name = 3;
  }

  repeated FeatureFlag feature_flags = 1;
}

message LinkFeatureFlagRequest {
  int64 warehouse_id = 1 [(validate.rules).int64.gt = 0];
  int64 feature_flag_id = 2 [(validate.rules).int64.gt = 0];
  int64 user_id = 3 [(validate.rules).int64.gt = 0];
}

message LinkFeatureFlagResponse {}

message UnlinkFeatureFlagRequest {
  int64 warehouse_id = 1 [(validate.rules).int64.gt = 0];
  int64 feature_flag_id = 2 [(validate.rules).int64.gt = 0];
  int64 user_id = 3 [(validate.rules).int64.gt = 0];
}

message UnlinkFeatureFlagResponse {}

message V1GetSplitRestrictionsRequest {
  int64 warehouse_id = 1 [(validate.rules).int64.gt = 0];
}

message V1GetSplitRestrictionsResponse {
  int64 warehouse_id = 1;
  google.protobuf.DoubleValue posting_max_volume = 2;
  google.protobuf.DoubleValue posting_max_weight_gr = 3;
  google.protobuf.DoubleValue item_max_volume = 4;
  google.protobuf.DoubleValue item_max_weight_gr = 5;
  google.protobuf.Int64Value item_max_dimension_mm = 6;
  repeated int64 item_single_instance_tags = 7;
  google.protobuf.Int64Value max_qty = 8;
}

message V1CreateSplitRestrictionsRequest {
  int64 warehouse_id = 1 [(validate.rules).int64.gt = 0];
  google.protobuf.DoubleValue posting_max_volume = 2;
  google.protobuf.DoubleValue posting_max_weight_gr = 3;
  google.protobuf.DoubleValue item_max_volume = 4;
  google.protobuf.DoubleValue item_max_weight_gr = 5;
  google.protobuf.Int64Value item_max_dimension_mm = 6;
  google.protobuf.Int64Value user_id = 7;
  repeated int64 item_single_instance_tags = 8 [(validate.rules).repeated = {unique: true, items: {int64: {gt: 0}}}];
  google.protobuf.Int64Value max_qty = 9;
  repeated int64 feature_flag_ids = 10 [(validate.rules).repeated = {unique: true, items: {int64: {gt: 0}}}];
}

message V1UpdateSplitRestrictionsRequest {
  int64 warehouse_id = 1 [(validate.rules).int64.gt = 0];
  google.protobuf.DoubleValue posting_max_volume = 2;
  google.protobuf.DoubleValue posting_max_weight_gr = 3;
  google.protobuf.DoubleValue item_max_volume = 4;
  google.protobuf.DoubleValue item_max_weight_gr = 5;
  google.protobuf.Int64Value item_max_dimension_mm = 6;
  google.protobuf.Int64Value user_id = 7;
  repeated int64 item_single_instance_tags = 8 [(validate.rules).repeated = {unique: true, items: {int64: {gt: 0}}}];
  google.protobuf.Int64Value max_qty = 9;
  repeated int64 feature_flag_ids = 10 [(validate.rules).repeated = {unique: true, items: {int64: {gt: 0}}}];
}

message V1CreateSplitRestrictionsResponse{}

message V1UpdateSplitRestrictionsResponse{}
